### PR TITLE
[tune] Dynamic Resources for Trials

### PR DESCRIPTION
**Pull request by richardliaw**
## What do these changes do?

Provides a small helper function for modifying the resource requirements of a trial.

Also implements the following:
 - setting the last_result to be {} instead of None
 - Adding a shuffle to the BasicVariantGenerator

### CODE REVIEW

File: python/ray/tune/schedulers/trial_scheduler.py [hunk 1]
```diff
@@ -11,6 +11,7 @@ class TrialScheduler(object):
     CONTINUE = "CONTINUE"  #: Status for continuing trial execution
     PAUSE = "PAUSE"  #: Status for pausing trial execution
     STOP = "STOP"  #: Status for stopping trial execution
+    NOOP = "NOOP"  #: Status for no-op (letting scheduler manage)
```
Comment (ericl) [inline python/ray/tune/schedulers/trial_scheduler.py]: I don't get this, the scheduler is returning NOOP, so who's making the decision then?

Comment (richardliaw) [reply python/ray/tune/schedulers/trial_scheduler.py]: The scheduler has access to the executor - for example if there are two trials, one of which you want to kill and the other you specifically want to subsume the resources of the first trial, returning "Trial.STOP" wouldn't work. Instead, all of this should be handled in the `scheduler.on_trial_result`, and the trial_runner doesn't have an explicit decision to make.

Comment (ericl) [reply python/ray/tune/schedulers/trial_scheduler.py]: The decision is CONTINUE isn't it?

The exact situation is handled already in PBT which resets trial params on the fly: https://github.com/ray-project/ray/blob/master/python/ray/tune/schedulers/pbt.py#L235

We could allow reset() to take new resources too.

Comment (richardliaw) [reply python/ray/tune/schedulers/trial_scheduler.py]: No. To clarify - the procedure is triggered by the return result of Trial 1; calling `CONTINUE` on Trial 1 (which you've killed) would result in error...

**Comment by ericl** [comment]
Why don't you just return STOP instead of killing it?

File: python/ray/tune/schedulers/trial_scheduler.py (hunk 1 above)
Comment (ericl) [reply python/ray/tune/schedulers/trial_scheduler.py]: Why don't you just return STOP instead of killing it?

**Comment by richardliaw** [comment]
Say all of your cluster is utilized:
- Trial 1 (1 CPU); Trial 2 (1 CPU) ... Trial N ...

Trial 1 returns a result, you now want:
- Trial 1 (0 CPU/Dead), Trial 2 (2 CPU) ... Trial N ...

If you to Trial.STOP, you need to do:
- Trial 1 (1 CPU); Trial 2 (1 CPU) ... Trial N ...
[resize Trial 2]
- Trial 1 (1 CPU); Trial 2 (2 CPU) ... Trial N ...
[return Trial.STOP]
- Trial 1 (0 CPU); Trial 2 (2 CPU) ... Trial N ...

But this leads oversubscription while trial 1 is waiting to be killed (and
is also much more annoying to program against, especially when readjusting
a batch of trials)

**Comment by ericl** [comment]
Yes, you should just allow it to be oversubscribed briefly. That's totally
fine and is handled by the Ray scheduler.

**Comment by richardliaw** [comment]
OK I see; removed the NOOP for now.

**Review by ericl — APPROVED** [review]

**Pull request merged**